### ISSUE TITLE
Mock expectation errors should be re-thrown when calling mock.verify()

### ISSUE THREAD

**Issue by DanReyLop** [sinonjs/sinon#1166]
* Sinon version : 2.0.0-pre.3 (`master` HEAD)
* Environment   : node 6.9.0
* Example URL   : N/A
* Other libraries you are using: N/A

**What did you expect to happen?**
A mock expectation error is thrown inside the code under test. Even if the exception is caight and silenced, I would expect that when calling `mock.verify()` the exception would be thrown again and make the test fail.

**What actually happens**
The `mock.verify()` call passes.

**How to reproduce**
I've made a small test that demonstrate this issue: [https://github.com/DanReyLop/sinon/commit/adf883215d5092844f6dcb63995deca72bafc37d](https://github.com/DanReyLop/sinon/commit/adf883215d5092844f6dcb63995deca72bafc37d)
In that test, a `mock` is expected to be called `once`, yet it's called `twice` (the second time, inside a `try` with an empty `catch` block) and the test passes. I know this is a contrived example, but it's not that uncommon to find something like this in a real-world application.
That test is included in the branch `verify-silenced-exceptions` of my fork ([here](https://github.com/DanReyLop/sinon/tree/verify-silenced-exceptions))

I'll be happy to submit a PR if you folks agree that's a good idea to tackle this issue.

**Comment by mroderick** [comment]
I agree with your analysis that `mock.verify()` should cause the test to fail, even when errors caused by calling `this.object.method()` are suppressed.

I would welcome a pull request that addresses this issue.

Here is a question to consider: what should happen if there are multiple failures suppressed? What would be the error thrown when calling `mock.verify()`?

```javascript
it("fails even if the original expectation exception was caught", function() {
    var mock = this.mock;
    mock.expects("method").once();

    this.object.method();

    for (var i = 0; i < 10; i++){
        try {
            this.object.method();
        } catch (e) {
            // Silenced error
        }        
    }

    assert.exception(function() {
        mock.verify();
    }, "ExpectationError");
});
```

Would there be details about all the failures in the `ExpecationError`? Or, will we say that re-throwing the first one is sufficient?

**Comment by DanReyLop** [comment]
I would say that re-throwing the first error is enough. Storing all the errors is technically easy, what I'm worried about is how to present them. In an example like yours, having it throw 10 stacktraces would be way too much. Maybe adding ` (and 9 more)` to the error message would be a good compromise?

**Comment by mroderick** [comment]
> Maybe adding (`and 9 more`) to the error message would be a good compromise?

I like this idea 👍
